MZ re-slotting code

joacoc opened the pull request:
<!--
Describe the contents of the PR briefly but completely.

If you write detailed commit messages, it is acceptable to copy/paste them
here, or write "see commit messages for details." If there is only one commit
in the PR, GitHub will have already added its commit message above.
-->

### Motivation

The new `mz` CLI Code contains a whole new code structure ([fork from here](https://github.com/benesch/materialize/tree/mz-refactor/src/mz/src)).

1.	The CLI is built in `clap` and using a similar command strucs as before, what mostly changes is the organization of the files and code.
2.	A context (`context.rs`) is built before every command executes. The context contains almost all the information the command needs, and they come in multiple layers:
	4. `Context`: Contains the configuration file and the output format.
	5. `ProfileContext`: Contains the `Context`, the profile information and clients (`Frontegg`, `Cloud` and `SQL`).
	6. `RegionContext`: Contains the `Profile Context` and the region information.
3.	Every command implementation divides into two parts:
	3. `/bin/command`: Contains the [Command] and [SubCommand] for `clap` and activates the context is activated. 
	4. `/command`: Contains the command and its subcommands implementation.

I'm still:
* Documenting part of the code.
* Fully matching features/output with [the design document.] (https://github.com/MaterializeInc/materialize/pull/17880)
* Polishing issues/features.
* Tests are still immature. 

<!--
Which of the following best describes the motivation behind this PR?

  * This PR fixes a recognized bug.

    [Ensure issue is linked somewhere.]

  * This PR adds a known-desirable feature.

    [Ensure issue is linked somewhere.]

  * This PR fixes a previously unreported bug.

    [Describe the bug in detail, as if you were filing a bug report.]

  * This PR adds a feature that has not yet been specified.

    [Write a brief specification for the feature, including justification
     for its inclusion in Materialize, as if you were writing the original
     feature specification.]

   * This PR refactors existing code.

    [Describe what was wrong with the existing code, if it is not obvious.]
-->

### Tips for reviewer

1. Profile (`profile.rs`) handling is by far the code's biggest and most influential part.
2. There are some `.unwrap()` around. Most of the times I replace the with custom errors, the idea before merging is to have none and have all of the documented in the `Error` struct.
3. I have two side PRs to merge to make this current code works.
    - [ ] Admin Client returning the user claims
    - [ ] Cloud Client accepting an `Arc<AdminClient>` to avoid having multiple clients around.    

<!--
Leave some tips for your reviewer, like:

    * The diff is much smaller if viewed with whitespace hidden.
    * [Some function/module/file] deserves extra attention.
    * [Some function/module/file] is pure code movement and only needs a skim.

Delete this section if no tips.
-->

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  <!-- Reference the design in the description. -->
- [ ] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  <!-- Ask in #team-cloud on Slack if you need help preparing the cloud PR. -->
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  - <!-- Add release notes here or explicitly state that there are no user-facing behavior changes. -->

benesch commented:
Amazing! This is a massive PR so it's going to take me a while to review. Hopefully will have some time this weekend.

sjwiesman commented on src/mz/src/server.rs:
```diff
@@ -0,0 +1,50 @@
+    Query(BrowserAPIToken { secret, client_id }): Query<BrowserAPIToken>,
+    tx: UnboundedSender<AppPassword>,
+) -> impl IntoResponse {
+    tx.send(AppPassword {
+        client_id: client_id.parse().unwrap(),
+        secret_key: secret.parse().unwrap(),
+    })
```
If the user clicks `no` on the web console to create an app password then the unwrap on line 29 is triggered. We should handle that case more gracefully.

sjwiesman replied:
It also hangs forever if you panic in this task. If anything fails you'll need to explicitly exit.

joacoc replied:
I've added multiple error handlers for the App Password. The only one missing is the transaction panicking, which has a low chance to occur but anyways we should have something there.

sjwiesman commented on src/mz/src/command/profile.rs:
```diff
@@ -0,0 +1,295 @@
+    let (server, port) = server(tx);
+
+    // Build the login URL
+    let mut url = DEFAULT_ENDPOINT.clone();
```
This function should take the hidden `cloud-endpoint` flag as a parameter to allow using mz against staging.

sjwiesman replied:
This works
```patch
diff --git a/src/mz/src/bin/mz/command/profile.rs b/src/mz/src/bin/mz/command/profile.rs
index 9fcff1289..b5b00a281 100644
--- a/src/mz/src/bin/mz/command/profile.rs
+++ b/src/mz/src/bin/mz/command/profile.rs
@@ -89,6 +89,7 @@ pub async fn run(mut cx: Context, cmd: ProfileCommand) -> Result<(), Error> {
                 &mut cx,
                 profile,
                 *no_browser,
+                endpoint.cloud_endpoint.clone(),
                 endpoint.admin_endpoint.clone(),
             )
             .await
diff --git a/src/mz/src/command/profile.rs b/src/mz/src/command/profile.rs
index 30cb716f2..50a181f8e 100644
--- a/src/mz/src/command/profile.rs
+++ b/src/mz/src/command/profile.rs
@@ -42,13 +42,13 @@ use crate::{
 
 /// Opens the default web browser in the host machine
 /// and awaits a single request containing the profile's app password.
-pub async fn init_with_browser() -> Result<AppPassword, Error> {
+pub async fn init_with_browser(cloud_endpoint: Option<Url>) -> Result<AppPassword, Error> {
     // Bind a web server to a local port to receive the app password.
     let (tx, mut rx) = mpsc::unbounded_channel();
     let (server, port) = server(tx);
 
     // Build the login URL
-    let mut url = DEFAULT_ENDPOINT.clone();
+    let mut url = cloud_endpoint.unwrap_or_else(|| DEFAULT_ENDPOINT.clone());
     url.path_segments_mut()
         .expect("constructor validated URL can be a base")
         .extend(&["account", "login"]);
@@ -136,11 +136,12 @@ pub async fn init(
     scx: &mut Context,
     profile_name: Option<String>,
     no_browser: bool,
+    cloud_endpoint: Option<Url>,
     admin_endpoint: Option<Url>,
 ) -> Result<(), Error> {
     let app_password = match no_browser {
         true => init_without_browser(admin_endpoint.clone()).await?,
-        false => init_with_browser().await?,
+        false => init_with_browser(cloud_endpoint.clone()).await?,
     };
 
     let new_profile = TomlProfile {
@@ -148,7 +149,7 @@ pub async fn init(
         vault: None,
         region: None,
         admin_endpoint: admin_endpoint.map(|url| url.to_string()),
-        cloud_endpoint: None,
+        cloud_endpoint: cloud_endpoint.map(|url| url.to_string()),
     };
 
     scx.config_file()
diff --git a/src/mz/src/command/region.rs b/src/mz/src/command/region.rs
index d53679d48..76d19f56c 100644
--- a/src/mz/src/command/region.rs
+++ b/src/mz/src/command/region.rs
@@ -24,7 +24,7 @@ use crate::{context::RegionContext, error::Error};
 use mz_cloud_api::client::{cloud_provider::CloudProvider, environment::Environment};
 use serde::{Deserialize, Serialize};
 use tabled::Tabled;
-use tokio::time::{sleep};
+use tokio::time::sleep;
 
 /// Enable a region in the profile organization.
 ///
@@ -52,14 +52,14 @@ pub async fn enable(cx: RegionContext) -> Result<(), Error> {
         match cx.get_environment(region.clone()).await {
             Ok(environment) => {
                 if environment.resolvable {
-                    break Ok(environment)
+                    break Ok(environment);
                 }
-            },
+            }
             Err(e) => {
                 if tries == 10 {
                     break Err(e);
                 }
-            },
+            }
         }
         sleep(Duration::from_secs(10)).await;
     }?;
```

benesch commented:
This is looking good! I'm sorry it's taken me so long to get to this. I pushed up some minor refactorings to get past some formatting lints. This PR is huge, so I think we should aim to get it merged as soon as possible, and then we can iterate on it with follow up PRs in main. The good news is that there's enough code here that multiple people should be able to make improvements in parallel once it's on main.

@joacoc — what's the best course of action for the tests? It seems like they're failing due to a missing `CI_PASSWORD` environment variable. I'd recommend swapping that `env!` macro for a call to `std::env::var`, so that we don't fail to compile when the env var is missing, and then leaving the env var unset in CI for now. We can figure out how to get these running in CI in a future PR.

joacoc commented:
@benesch Tests are now split into two functions:
* `test_local` that works locally.
* `test_e2e [ignored]` that interacts with external services, like the Cloud, Frontegg, etc.

As a first step I want to pass all the checks. After that, as the next step I have two options:

1. Finding the way to get the env variables missing working. (I've added `CLOUD_ENDPOINT` and `ADMIN_ENDPOINT`)
2. Mocking as much as I can without messing with the CI process (copying the [mock server available in environmentd](https://github.com/MaterializeInc/materialize/blob/3d742bdb39dba17b8b738aafccce23771b5376b8/src/environmentd/tests/sql.rs#L120)).

benesch commented on src/mz/tests/e2e.rs:
```diff
@@ -0,0 +1,537 @@
+        assert!(output.trim() == expected_command_output.trim());
+    }
+
+    /// TODO: Re-enable after understanding how to get an CI app-password, admin endpoint and cloud endpoint.
```
Do we have a tracking issue for `mz` somewhere? If not, we should file one, and make sure this is listed as a blocker for public preview!

benesch reviewed (approved):
I wasn't able to review the code in much detail, but that's okay—we can iterate on this as we go! The basic structure here looks really good, and I'm excited to get this into people's hands. 

LGTM, but please be sure to squash merge to keep the history on main clean.

joacoc commented on test/cloud-canary/mzcompose.py:
```diff
@@ -88,7 +88,7 @@ def workflow_default(c: Composition, parser: WorkflowArgumentParser) -> None:
     try:
         print(f"Enabling region using Mz version {VERSION} ...")
         try:
-            c.run("mz", "region", "enable", REGION, "--version", VERSION)
```
Correct

philip-stoev commented on misc/python/materialize/mzcompose/services.py:
```diff
@@ -1050,13 +1049,13 @@ def __init__(
+                "profile": "default",
                 "profiles": {
                     "default": {
-                        "email": username,
```
So authentication no longer requires a username, just the app_password?

joacoc replied:
`app_password` is enough. The CLI retrieves the username from Frontegg.
